### PR TITLE
Generate images file that includes sources

### PR DESCRIPTION
**Pull request by rmweir**
**Problem:**
Prior, a list of images was produced but there was no annotation included for why each image is included. This information is useful in deducing whether an airgap setup absolutely needs an image.

**Solution:**
Image export now produces a text file that includes images and which part of rancher utilizes them. Possible image sources are any feature or system chart, rke, k3sUpgrade, core (meaning it is need for rancher core functionality), and rancher (which it is either the rancher server image or the agent image).

**Testing:**
To test compare rancher-images.txt to previous version and make sure there aren't any unexpected difference. An unexpected difference would be something like many images missing. Another thing to validate is the the rancher-images-sources.txt should contain the same number of rows as the rancher-images.txt.

**Issue:**
https://github.com/rancher/rancher/issues/31663

**Related Issue:**
https://github.com/rancher/rancher/issues/29893

### CODE REVIEW

File: pkg/image/resolve.go [hunk 1]
```diff
@@ -240,38 +278,41 @@ func fetchImagesFromCollection(obj map[string]interface{}) (images []string) {
...
+
+func getSourcesList(sources map[string]bool) string {
+	commaSeparatedSources := ""
+	for source, val := range sources {
+		if !val {
+			continue
+		}
+		commaSeparatedSources += fmt.Sprintf(",%s", source)
+	}
+	return strings.TrimPrefix(commaSeparatedSources, ",")
```
Comment (StrongMonkey) [inline pkg/image/resolve.go:317]:
```suggestion
	var commaSeparatedSources []string
	for source, val := range sources {
		if !val {
			continue
		}
		commaSeparatedSources = append(commaSeparatedSources, source)
	}
	return strings.Join(commaSeparatedSources, ",")

```

**Review by StrongMonkey — COMMENTED** [review]
LGTM except one NIT. @rmweir Can you also paste the image run output in the comment we can have a better idea?

File: pkg/image/resolve.go (hunk 1 above)
Comment (rmweir) [reply pkg/image/resolve.go:317]: sure, here is an excerpt of the new rancher-images-sources.txt:
```
rancher/externalip-webhook:v0.1.6 rancher-external-ip-webhook:0.1.601-rc01
rancher/flannel-cni:v0.3.0-rancher6 rke
rancher/fleet-agent:v0.3.4-rc7 fleet:0.3.400-rc07,fleet-agent:0.3.400-rc07
rancher/fleet:v0.3.4-rc7 fleet:0.3.400-rc07
rancher/fluentd:v0.1.24 rancher-logging:0.2.4
rancher/gitjob:v0.1.13 fleet:0.3.400-rc07
rancher/gke-operator:v1.0.1-rc1 rancher-gke-operator:1.0.100
rancher/grafana-grafana:6.7.4 rancher-istio:1.5.901
rancher/grafana-grafana:7.1.5 rancher-monitoring:0.2.1
rancher/hyperkube:v1.16.15-rancher1 rke
```

Comment (StrongMonkey) [reply pkg/image/resolve.go:317]: ok thank you. I think if we can sort based on source it would look better? so that images are grouped by their source

**Review by StrongMonkey — APPROVED** [review]
:rocket:

**Review by brendarearden — APPROVED** [review]
LGTM

**Review by StrongMonkey — APPROVED** [review]

**Review by brendarearden — APPROVED** [review]

**Comment by rmweir** [comment]
@brendarearden @StrongMonkey updated so that sources order in rancher-images-sources.txt are deterministic.

**Review by StrongMonkey — APPROVED** [review]

**Review by brendarearden — APPROVED** [review]

File: pkg/image/resolve.go [hunk 2]
```diff
@@ -132,79 +154,86 @@ func walkthroughMap(inputMap map[interface{}]interface{}, walkFunc func(map[inte
...
+		if err := fetchImagesFromSystem(rkeSystemImages, osType, imagesSet); err != nil {
+			return nil, nil, errors.Wrap(err, "failed to fetch images from system images")
 		}
-		images = append(images, imagesInSystem...)
 	}
 
-	// append images from requirement
-	if requirementImages := getRequirementImages(osType); len(requirementImages) > 0 {
-		images = append(images, requirementImages...)
-	}
```
Comment (cbron) [inline pkg/image/resolve.go:168]: why arent we checking length of requirementImages anymore ? Is it because they get checked in `addSourceToImage` ?

**Review by cbron — COMMENTED** [review]
Seems fine. Make sure qa tests old list vs this one in an RC.

File: pkg/image/resolve.go (hunk 2 above)
Comment (rmweir) [reply pkg/image/resolve.go:168]: It no longer returns a slice, it just sets the images in the map if applicable.

**Review by cbron — APPROVED** [review]

**Comment by cbron** [comment]
Needs to get forwardported also.

**Pull request merged**